### PR TITLE
Removed default values to fix EDRS configuration error for 1NODE SDDC

### LINKED ISSUE #76: edrs settings are not truly "Optional"
It appears that the new `edrs` settings are not truly optional.

This manifests when using a single node (1NODE) type SDDC.

Using a terraform config where there are **no** edrs settings defined, we get an error on an update run.

The initial create works fine for some reason, however a subsequent plan and apply tries to change the edrs settings.  This then fails, because it tries to change them to the default which are invalid for 1NODE type.

Change seen in plan:
```
      ~ edrs_policy_type       = "cost" -> "storage-scaleup"                                                                                      
      ~ enable_edrs            = false -> true 
```

Error seen on apply:
```
Error:  Failed to update EDRS Policy: [Elastic DRS is not supported for sddc ff71f299-d38c-453d-9b19-97f2407f9f35, sddcType 1NODE, deployment type
 SingleAZ. (code)115 ]                                                                                                                            
                                                                                                                                                  
  on module/SDDC/main.tf line 1, in resource "vmc_sddc" "this":                                                                                   
   1: resource "vmc_sddc" "this" {                                                                                                                                                                                                                                                  
```

### LINKED ISSUE #79: min_hosts key:value shows as a change on each apply when using Stretched SDDC
When run plans/applies against a stretched sddc post build I am seeing  ~ min_hosts   = 6 -> 3  causing the output that the sddc is to be changed rather than everything is up to date.

### LINKED ISSUE #82: Provider does not detect failed SDDC create, waits forever
If an SDDC fails to create, the `vmc_sddc` resource will wait forever for the creation to succeed, instead of reporting the error.

Screenshot of SDDC in failed state:
image

Our Terraform `create` timeout is set to 300 minutes.  Our pipeline thus finished at the 5 hour mark below, instead of finishing as soon as the SDDC entered FAILED state:

Log output:
```
...
module.sddc-1.vmc_sddc.this: Still creating... [4h59m50s elapsed]
module.sddc-1.vmc_sddc.this: Still creating... [5h0m0s elapsed]
                                                                                                                                                  
Error: expected instance to be created but was in state FAILED 
                                                                                                                                                  
  on module/SDDC/main.tf line 1, in resource "vmc_sddc" "this":                      
   1: resource "vmc_sddc" "this" {                                                             

```

### PR DESCRIPTION
**Pull request by saloneerege**
Added error messaged to be returned when the task returns a FAILED state for create, update and delete operations

### CODE REVIEW

File: vmc/resource_vmc_cluster.go [hunk 1]
```diff
@@ -215,6 +215,9 @@ func resourceClusterCreate(d *schema.ResourceData, m interface{}) error {
 			}
 			d.SetId(clusterID)
 		}
+		if *task.Status == "FAILED" {
```
Comment (rchadalawada) [inline vmc/resource_vmc_cluster.go:218]: I think the error messages could be reworded. Something like... "Task failed to create instance" & "Task completed without creating instance"

File: vmc/resource_vmc_cluster.go [hunk 2]
```diff
@@ -256,8 +259,10 @@ func resourceClusterRead(d *schema.ResourceData, m interface{}) error {
 			cluster["cluster_name"] = *clusterConfig.ClusterName
 			cluster["cluster_state"] = *clusterConfig.ClusterState
 			cluster["host_instance_type"] = *clusterConfig.EsxHostInfo.InstanceType
-			cluster["mssql_licensing"] = *clusterConfig.MsftLicenseConfig.MssqlLicensing
-			cluster["windows_licensing"] = *clusterConfig.MsftLicenseConfig.WindowsLicensing
+			if clusterConfig.MsftLicenseConfig != nil {
```
Comment (rchadalawada) [inline vmc/resource_vmc_cluster.go:261]: I suppose this check is not related to the fix

Comment (saloneerege) [reply vmc/resource_vmc_cluster.go:261]: That is correct. While testing this fix, I came across a nil value being returned on resource creation when msft licensing config is not provided in the resource creation.

File: vmc/resource_vmc_cluster.go (hunk 1 above)
Comment (saloneerege) [reply vmc/resource_vmc_cluster.go:218]: Sure.

File: vmc/constants.go [hunk 3]
```diff
@@ -48,6 +48,9 @@ const (
 	CapitalLicenseConfigEnabled  = "ENABLED"
 	CapitalLicenseConfigDisabled = "DISABLED"
 
+	//SDDC Type
```
Comment (shylasrinivas) [inline vmc/constants.go]: space after //

File: vmc/resource_vmc_cluster.go (hunk 1 above)
Comment (shylasrinivas) [reply vmc/resource_vmc_cluster.go:218]: these errors are displayed in TF console. Is it required to explicitly state task in the message  to a TF user?

File: vmc/resource_vmc_cluster.go [hunk 4]
```diff
@@ -293,6 +298,9 @@ func resourceClusterDelete(d *schema.ResourceData, m interface{}) error {
 		if err != nil {
 			return resource.NonRetryableError(fmt.Errorf("error while deleting SDDC %s: %v", sddcID, err))
 		}
+		if *task.Status == "FAILED" {
```
Comment (shylasrinivas) [inline vmc/resource_vmc_cluster.go:300]: can FAILED and FINISHED (line 304) be combined into one if with ||? I don't see a difference in the way they are both handled.

File: vmc/resource_vmc_cluster.go [hunk 5]
```diff
@@ -338,6 +346,9 @@ func resourceClusterUpdate(d *schema.ResourceData, m interface{}) error {
 			if err != nil {
 				return resource.NonRetryableError(fmt.Errorf("error while waiting for task %s: %v", task.Id, err))
 			}
+			if *task.Status == "FAILED" {
```
Comment (shylasrinivas) [inline vmc/resource_vmc_cluster.go:347]: same comment as above

File: vmc/resource_vmc_sddc.go [hunk 6]
```diff
@@ -367,6 +363,9 @@ func resourceSddcCreate(d *schema.ResourceData, m interface{}) error {
 			return resource.NonRetryableError(fmt.Errorf("error describing instance: %s", err))
 
 		}
+		if *task.Status == "FAILED" {
```
Comment (shylasrinivas) [inline vmc/resource_vmc_sddc.go:366]: comible FAILED and FINISHED into one if

**Review by shylasrinivas — CHANGES_REQUESTED** [review]

File: vmc/constants.go (hunk 3 above)
Comment (saloneerege) [reply vmc/constants.go]: Done

File: vmc/resource_vmc_cluster.go (hunk 2 above)
Comment (saloneerege) [reply vmc/resource_vmc_cluster.go:261]: Done

File: vmc/resource_vmc_cluster.go [hunk 7]
```diff
@@ -197,26 +197,28 @@ func resourceClusterCreate(d *schema.ResourceData, m interface{}) error {
...
 
 			}
 			d.SetId(clusterID)
 		}
-		if *task.Status != "FINISHED" {
-			return resource.RetryableError(fmt.Errorf("expected instance to be created but was in state %s", *task.Status))
+		if *task.Status == "FAILED" {
+			return resource.NonRetryableError(fmt.Errorf("task failed to create cluster"))
+		} else if *task.Status != "FINISHED" {
+			return resource.RetryableError(fmt.Errorf("expected cluster to be create but was in state %s", *task.Status))
```
Comment (shylasrinivas) [inline vmc/resource_vmc_cluster.go]: typo create => created

File: vmc/resource_vmc_cluster.go [hunk 8]
```diff
@@ -291,10 +295,12 @@ func resourceClusterDelete(d *schema.ResourceData, m interface{}) error {
 	return resource.Retry(d.Timeout(schema.TimeoutDelete), func() *resource.RetryError {
 		task, err := tasksClient.Get(orgID, task.Id)
 		if err != nil {
-			return resource.NonRetryableError(fmt.Errorf("error while deleting SDDC %s: %v", sddcID, err))
+			return resource.NonRetryableError(fmt.Errorf("error deleting cluster %s : %v", clusterID, err))
 		}
-		if *task.Status != "FINISHED" {
-			return resource.RetryableError(fmt.Errorf("expected instance to be deleted but was in state %s", *task.Status))
+		if *task.Status == "FAILED" {
+			return resource.NonRetryableError(fmt.Errorf("task failed to create cluster"))
```
Comment (shylasrinivas) [inline vmc/resource_vmc_cluster.go]: supposed to be delete?

File: vmc/resource_vmc_cluster.go [hunk 9]
```diff
@@ -291,10 +295,12 @@ func resourceClusterDelete(d *schema.ResourceData, m interface{}) error {
...
+		} else if *task.Status != "FINISHED" {
+			return resource.RetryableError(fmt.Errorf("expected cluster to be created but was in state %s", *task.Status))
```
Comment (shylasrinivas) [inline vmc/resource_vmc_cluster.go]: created => deleted?

File: vmc/resource_vmc_cluster.go [hunk 10]
```diff
@@ -368,18 +379,19 @@ func resourceClusterUpdate(d *schema.ResourceData, m interface{}) error {
...
 					return resource.RetryableError(fmt.Errorf("instance update still in progress"))
 				}
-				return resource.NonRetryableError(fmt.Errorf("error describing instance: %s", err))
-
+				return resource.NonRetryableError(fmt.Errorf("error updating EDRS policy configuration : %v", err))
 			}
-			if *task.Status != "FINISHED" {
-				return resource.RetryableError(fmt.Errorf("expected instance to be updated but was in state %s", *task.Status))
+			if *task.Status == "FAILED" {
+				return resource.NonRetryableError(fmt.Errorf("task failed to EDRS policy configuration"))
```
Comment (shylasrinivas) [inline vmc/resource_vmc_cluster.go]: an action verb is required in this message Eg: update

File: vmc/resource_vmc_cluster.go [hunk 11]
```diff
@@ -397,18 +409,19 @@ func resourceClusterUpdate(d *schema.ResourceData, m interface{}) error {
...
 					return resource.RetryableError(fmt.Errorf("instance update still in progress"))
 				}
-				return resource.NonRetryableError(fmt.Errorf("error describing instance: %s", err))
-
+				return resource.NonRetryableError(fmt.Errorf("error updating microsoft licensing configuration : %v", err))
 			}
-			if *task.Status != "FINISHED" {
-				return resource.RetryableError(fmt.Errorf("expected instance to be updated but was in state %s", *task.Status))
+			if *task.Status == "FAILED" {
+				return resource.NonRetryableError(fmt.Errorf("task failed to microsoft licensing configuration"))
```
Comment (shylasrinivas) [inline vmc/resource_vmc_cluster.go]: same as above.. update in message

File: vmc/resource_vmc_sddc.go [hunk 12]
```diff
@@ -596,18 +601,26 @@ func resourceSddcUpdate(d *schema.ResourceData, m interface{}) error {
 	}
 
 	if d.HasChange("edrs_policy_type") || d.HasChange("enable_edrs") || d.HasChange("min_hosts") || d.HasChange("max_hosts") {
-		edrsPolicyClient := autoscalercluster.NewDefaultEdrsPolicyClient(connectorWrapper)
+
+		SDDCType := d.Get("sddc_type").(string)
```
Comment (shylasrinivas) [inline vmc/resource_vmc_sddc.go]: Is SDDCType a constant or exported variable? Wondering why this naming convention is required..

**Review by shylasrinivas — CHANGES_REQUESTED** [review]
minor typos

File: vmc/resource_vmc_cluster.go (hunk 7 above)
Comment (saloneerege) [reply vmc/resource_vmc_cluster.go]: Done

File: vmc/resource_vmc_cluster.go (hunk 8 above)
Comment (saloneerege) [reply vmc/resource_vmc_cluster.go]: Done

File: vmc/resource_vmc_cluster.go (hunk 9 above)
Comment (saloneerege) [reply vmc/resource_vmc_cluster.go]: Done

File: vmc/resource_vmc_cluster.go (hunk 10 above)
Comment (saloneerege) [reply vmc/resource_vmc_cluster.go]: Done

File: vmc/resource_vmc_cluster.go (hunk 11 above)
Comment (saloneerege) [reply vmc/resource_vmc_cluster.go]: Done

**Review by shylasrinivas — APPROVED** [review]
lgtm

**Review by shylasrinivas — APPROVED** [review]

**Comment by carsonpadawon** [comment]
Trying to update the SDDC with network segments, but I get the same error listed above.  
image
Here is the version I am running
image

**Pull request merged**